packages/pigeon/lib/dart_generator.dart

```diff
@@ -521,92 +436,1264 @@ final BinaryMessenger? ${_varNamePrefix}binaryMessenger;
+                _writeHostMethodMessageCall(
+                  Indent(messageCallSink),
+                  channelName: makeChannelNameWithStrings(
+                    apiName: apiName,
+                    methodName: method.name,
+                    dartPackageName: dartPackageName,
+                  ),
+                  parameters: <Parameter>[
+                    if (!method.isStatic)
+                      Parameter(
+                        name: 'this',
+                        type: TypeDeclaration(
+                          baseName: apiName,
+                          isNullable: false,
+                        ),
+                      ),
+                    ...method.parameters,
+                  ],
+                  returnType: method.returnType,
+                );
+                builder.statements.addAll(<cb.Code>[
+                  if (!method.isStatic)
+                    cb.Code('final $codecName $_pigeonChannelCodec =\n'
+                        '    $codecInstanceName;')
+                  else
+                    cb.Code(
+                      'final $codecName $_pigeonChannelCodec = $codecName(${classMemberNamePrefix}instanceManager ?? $instanceManagerClassName.instance);',
+                    ),
+                  const cb.Code(
+                    'final BinaryMessenger? ${_varNamePrefix}binaryMessenger = ${classMemberNamePrefix}binaryMessenger;',
+                  ),
+                  cb.Code(messageCallSink.toString()),
+                ]);
+              },
+            ),
+        ),
+      cb.Method(
+        (cb.MethodBuilder builder) => builder
+          ..name = '${classMemberNamePrefix}copy'
+          ..returns = cb.refer(apiName)
+          ..annotations.add(cb.refer('override'))
+          ..body = cb.Block.of(<cb.Code>[
+            cb
+                .refer('$apiName.${classMemberNamePrefix}detached')
+                .call(
+                  <cb.Expression>[],
+                  <String, cb.Expression>{
+                    '${classMemberNamePrefix}binaryMessenger':
+                        cb.refer('${classMemberNamePrefix}binaryMessenger'),
+                    '${classMemberNamePrefix}instanceManager':
+                        cb.refer('${classMemberNamePrefix}instanceManager'),
+                    for (final Field field in unattachedFields)
+                      field.name: cb.refer(field.name),
+                    for (final Method method in superClassFlutterMethods)
+                      method.name: cb.refer(method.name),
+                    for (final AstProxyApi proxyApi in interfacesApis)
+                      for (final Method method in proxyApi.methods)
+                        method.name: cb.refer(method.name),
+                    for (final Method method in flutterMethods)
+                      method.name: cb.refer(method.name),
+                  },
+                )
+                .returned
+                .statement,
+          ]),
+      ),
+    ];
+  }
```

Comment:
This is quite a method.

Maybe we could have a separate file for all the proxy-specific code? Then it would be easier to do things like extract helpers for different sections without overwhelming the rest of this file.